src/topology/algebra/uniform_ring.lean

```diff
@@ -150,6 +151,31 @@ instance : comm_ring (completion R) :=
       (assume a b, by rw [← coe_mul, ← coe_mul, mul_comm]),
  ..completion.ring }
 
+instance ring_completion.algebra : algebra R (completion R) :=
+(uniform_space.completion.coe_ring_hom : R →+* (completion R)).to_algebra
+
+lemma ring_completion.algebra_smul_eq (r : R) (x : (completion R)) :
+  r • x = (r : completion R) * x :=
+rfl
+
+section algebra
+variables (S : Type*) [comm_ring S] [algebra S R]
+
+instance ring_completion.algebra' : algebra S (completion R) :=
+((uniform_space.completion.coe_ring_hom : R →+* completion R).comp (algebra_map S R)).to_algebra
+
```

Comment:
This instance has the potential to form algebra diamonds, for example:
```lean
example : (algebra_int (completion R)).smul = (ring_completion.algebra' R ℤ).smul := rfl -- fails
```
And the same for `ℕ` if we weaken the `comm_ring S` assumption above. I can't tell for sure whether you're likely to multiply integers with `completion R` (I imagine you might want to but the diamond has been lurking for longer so maybe not), but it's likely to be very painful if you do.

To fix the diamond, you should supply a custom definition of the `nsmul` and `zmul` fields for the `add_monoid`..`comm_group` instances on `completion R`, so that `n \bu x` first casts `n` to `R` before multiplying.

Reply:
I've made an attempt at filling those fields in #14846, but using a different definition to the one you suggest.

Reply:
Thanks for doing this! I'll try to incorporate it now.